cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp

```diff
@@ -724,15 +724,15 @@ void iterate(const std::vector<T>& v) {
 }
 
 std::vector<int>& ref_to_first_in_returnValue_1() {
-  return returnValue()[0]; // BAD [NOT DETECTED] (see *)
+  return returnValue()[0]; // BAD
 }
 
 std::vector<int>& ref_to_first_in_returnValue_2() {
   return returnValue()[0]; // BAD [NOT DETECTED]
```

Comment:
Why is this case not detected? The function signature and implementation is identical to `ref_to_first_in_returnValue_1`

Reply:
I haven't investigated why yet. Note that, while function signature and implementation is identical to `ref_to_first_in_returnValue_1`, they differ in how they're used. See [here](https://github.com/github/codeql/blob/5f8eb7b1381eea0d75924171c58d6e0530b22784/cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp#L750-L755) for this.

In the one we detect (i.e., `ref_to_first_in_returnValue_1`) the range-based for loop looks like:
```cpp
for (auto x : ref_to_first_in_returnValue_1()) {}
```
and in the one we fail to detect the range-based for loop looks like:
```cpp
{
auto value = ref_to_first_in_returnValue_2();
for (auto x : value) {}
}
```